src/audio/tdfb/tdfb.c

```diff
@@ -79,6 +79,187 @@ static inline int set_func(struct processing_module *mod, enum sof_ipc_frame fmt
+		f = MIN(f, fmax);
+		for (j = 0; j < f; j += 2) {
+			for (i = 0; i < num_filters; i++) {
+				is = cd->input_channel_select[i];
+				om = cd->output_channel_mix[i];
+				s0 = x[is];
+				s1 = x[is + in_nch];
+				for (k = 0; k < out_nch; k++) {
+					if (om & 1) {
+						y[k] = s0;
+						y[k + out_nch] = s1;
+					}
+					om = om >> 1;
+				}
+			}
+			x += 2 * in_nch;
+			y += 2 * out_nch;
+		}
+		remaining_frames -= f;
+		x = audio_stream_wrap(source, x);
+		y = audio_stream_wrap(sink, y);
+	}
+}
+#endif
+
+#if CONFIG_FORMAT_S24LE
+static void tdfb_pass_s24(struct tdfb_comp_data *cd, struct input_stream_buffer *bsource,
+			  struct output_stream_buffer *bsink, int frames)
+{
+	struct audio_stream *source = bsource->data;
+	struct audio_stream *sink = bsink->data;
+	int32_t *x = audio_stream_get_rptr(source);
+	int32_t *y = audio_stream_get_wptr(sink);
+	int32_t s0, s1;
+	int fmax;
+	int is, om;
+	int f, i, j, k;
+	const int in_nch = audio_stream_get_channels(source);
+	const int out_nch = audio_stream_get_channels(sink);
+	const int num_filters = cd->config->num_filters;
+	int remaining_frames = frames;
+
+	while (remaining_frames) {
+		fmax = audio_stream_frames_without_wrap(source, x);
+		f = MIN(remaining_frames, fmax);
+		fmax = audio_stream_frames_without_wrap(sink, y);
+		f = MIN(f, fmax);
+		for (j = 0; j < f; j += 2) {
+			for (i = 0; i < num_filters; i++) {
+				is = cd->input_channel_select[i];
+				om = cd->output_channel_mix[i];
+				s0 = x[is];
+				s1 = x[is + in_nch];
+				for (k = 0; k < out_nch; k++) {
+					if (om & 1) {
+						y[k] = s0;
+						y[k + out_nch] = s1;
+					}
+					om = om >> 1;
+				}
+			}
+			x += 2 * in_nch;
+			y += 2 * out_nch;
+		}
+		remaining_frames -= f;
+		x = audio_stream_wrap(source, x);
+		y = audio_stream_wrap(sink, y);
+	}
+}
+#endif
+
+#if CONFIG_FORMAT_S32LE
+static void tdfb_pass_s32(struct tdfb_comp_data *cd, struct input_stream_buffer *bsource,
+			  struct output_stream_buffer *bsink, int frames)
+{
+	struct audio_stream *source = bsource->data;
+	struct audio_stream *sink = bsink->data;
+	int32_t *x = audio_stream_get_rptr(source);
+	int32_t *y = audio_stream_get_wptr(sink);
+	int32_t s0, s1;
+	int fmax;
+	int is, om;
+	int f, i, j, k;
+	const int in_nch = audio_stream_get_channels(source);
+	const int out_nch = audio_stream_get_channels(sink);
+	const int num_filters = cd->config->num_filters;
+	int remaining_frames = frames;
+
+	while (remaining_frames) {
+		fmax = audio_stream_frames_without_wrap(source, x);
+		f = MIN(remaining_frames, fmax);
+		fmax = audio_stream_frames_without_wrap(sink, y);
+		f = MIN(f, fmax);
+		for (j = 0; j < f; j += 2) {
+			for (i = 0; i < num_filters; i++) {
+				is = cd->input_channel_select[i];
+				om = cd->output_channel_mix[i];
+				s0 = x[is];
+				s1 = x[is + in_nch];
+				for (k = 0; k < out_nch; k++) {
+					if (om & 1) {
+						y[k] = s0;
+						y[k + out_nch] = s1;
+					}
+					om = om >> 1;
```

Comment:
in theory adding a
```
if (!om)
    break;
```
could save some time in some cases, but that's probably rare? More often than not you'd waste time on this test instead of saving it?

Reply:
Yes, I'm pretty sure it would not save cycles.